miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,385 @@
+        let SparseMerklePath { empty_nodes, nodes } = self;
+        (empty_nodes, nodes)
+    }
+
+    /// Returns the total depth of this path, i.e., the number of nodes this path represents.
+    pub fn depth(&self) -> u8 {
+        (self.nodes.len() + self.empty_nodes.count_ones() as usize) as u8
+    }
+
+    /// Get a specific node in this path at a given depth.
+    ///
+    /// # Panics
+    /// With debug assertions enabled, this function panics if `node_depth` is greater than
+    /// `tree_depth` (as it is impossible to have a node of greater depth than the tree it is
+    /// contained in).
+    pub fn get(&self, node_depth: u8) -> Option<RpoDigest> {
+        if node_depth > self.depth() {
+            return None;
+        }
+
+        let empty_bit = u64::wrapping_shl(1, node_depth.into());
+        let is_empty = (self.empty_nodes & empty_bit) != 0;
+
+        if is_empty {
+            return Some(*EmptySubtreeRoots::entry(self.depth(), node_depth));
+        }
+
+        // Our index needs to account for all the empty nodes that aren't in `self.nodes`.
+        let nonempty_index: usize = {
+            // TODO: this could also be u64::unbounded_shl(1, node_depth + 1).wrapping_sub(1).
+            // We should check if that has any performance benefits over using 128-bit integers.
+            let mask: u64 = ((1u128 << (node_depth + 1)) - 1u128).try_into().unwrap();
+
+            let empty_before = u64::count_ones(self.empty_nodes & mask);
+            u64::checked_sub(node_depth as u64, empty_before as u64)
+                .unwrap()
+                .try_into()
+                .unwrap()
+        };
+        Some(self.nodes[nonempty_index])
+    }
+}
+
+// CONVERSIONS
+// ================================================================================================
+
+impl From<SparseMerklePath> for MerklePath {
+    fn from(sparse_path: SparseMerklePath) -> Self {
+        sparse_path.into_path()
+    }
+}
+
+/// # Panics
+///
+/// This conversion fails and panics if the path length is greater than [`SMT_MAX_DEPTH`].
+impl From<MerklePath> for SparseMerklePath {
+    fn from(path: MerklePath) -> Self {
+        SparseMerklePath::from_path(path).unwrap()
+    }
+}
+
+impl From<SparseMerklePath> for Vec<RpoDigest> {
+    fn from(path: SparseMerklePath) -> Self {
+        path.into_path().into()
+    }
+}
+
+// ITERATORS
+// ================================================================================================
+
+impl IntoIterator for SparseMerklePath {
+    type Item = <SparseMerkleIter as Iterator>::Item;
+    type IntoIter = SparseMerkleIter;
+
+    fn into_iter(self) -> SparseMerkleIter {
+        let tree_depth = self.depth();
+        SparseMerkleIter {
+            path: self,
+            next_depth: Some(0),
+            tree_depth,
+        }
+    }
+}
+
+/// Owning iterator for [`SparseMerklePath`].
+// TODO: add a non-owning iterator too.
+pub struct SparseMerkleIter {
```

Comment:
I think just to avoid confusion it would be best to name this `SparseMerklePathIter`.